# Sobre o notebook

A "hora da prática" é uma atividade do curso de Python para Data Science da Alura que contém diversos desafios envolvendo os aprendizados de cada aula.

O notebook "hora da pratica" é um notebook construído com espaço para as células dos desafios propostos no curso, com ele você tem espaço para construir e executar suas soluções

## Aula 1 - Introdução ao Python

### Questão 1

Imprima a frase `Escola de Dados da Alura!`

In [ ]:
print('Escola de Dados da Alura!')

Escola de Dados da Alura!


### Questão 2
Imprima na tela seu nome e seu sobrenome seguindo a estrutura abaixo:
```
Nome: [seu nome]
Sobrenome: [seu sobrenome]
```


In [ ]:
#Imprime os dados nome e sobrenome:
print('Nome: [Wender]')
print('Sobrenome: [Xavier]')

Nome: [Wender]
Sobrenome: [Xavier]


In [ ]:
#Imprime o nome, uma letra por linha:
print('W')
print('E')
print('N')
print('D')
print('E')
print('R')

W
E
N
D
E
R


### Questão 4
Imprima o dia do seu nascimento em formato `dia mês ano`. Os valores de dia e ano não podem estar entre aspas. O formato deve estar como no exemplo, supondo uma data de aniversário dia 28 de fevereiro de 2003:
```
28 fevereiro 2003
```

In [ ]:
#Imprime a data de nascimento:
print(3, 'outubro', 1992)

3 outubro 1992


### Questão 5
Imprima em um único print o atual ano que você está fazendo esse curso, o valor do ano deve ser um dado numérico. A saída do print deve ser a seguinte:
```
Ano atual: [ano]
```


In [ ]:
#Declara a variável ano e imprime o valor em seguida:
ano = 2023
print(f'Ano atual: {ano}')

Ano atual: 2023


## Aula 2 - Manipulando dados no Python

### Coleta e amostragem de dados

#### Questão 1
Crie um programa que solicite à pessoa usuária digitar seu nome, e imprima “Olá, [nome]!”.

In [ ]:
#Solicita ao usuário a entrada:
nome = str(input('Digite seu nome: '))
print(f'Olá, {nome}')# Imprime o valor digitado pelo usuário

Digite seu nome: João
Olá, João


#### Questão 2
Crie um programa que solicite à pessoa usuária digitar seu nome e idade, e imprima “Olá, [nome], você tem [idade] anos.”.

In [ ]:
name = str(input('Digite seu nome: '))
age = int(input('Qual é a sua idade?'))
print(f'Olá, {name}, você tem {age} anos!')

Digite seu nome: Julia
Qual é a sua idade?20
Olá, Julia, você tem 20 anos!


#### Questão 3
Crie um programa que solicite à pessoa usuária digitar seu nome, idade e altura em metros, e imprima “Olá, [nome], você tem [idade] anos e mede [altura] metros!”.

In [ ]:
#Importa o método replace
from os import replace
name = str(input('Digite seu nome: '))#Coleta nome
age = int(input('Qual é a sua idade?'))#Coleta idade
height = (input('Qual é a sua altura?')).replace(',', '.')#Coleta altura e converte , por . caso o usuário digite
print(f'Olá, {name}, você tem {age} anos e {height} de altura!')#Imprime os dados.

Digite seu nome: fef
Qual é a sua idade?13
Qual é a sua altura?1,56
Olá, fef, você tem 13 anos e 1.56 de altura!


### Calculadora com operadores

#### Questão 1

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a soma dos dois valores.

In [ ]:
#Coleta os valores e atribui às variáveis:
v1 = float(input('Digite um número: '))
v2 = float(input('Digite outro número: '))
soma = v1 + v2#Soma as duas variáveis
print(f'A soma de {v1} e {v2} é: {soma}')#Imprime o resultado

Digite um número: 23
Digite outro número: 56
A soma de 23.0 e 56.0 é: 79.0


#### Questão 2
Crie um programa que solicite três valores numéricos à pessoa usuária e imprima a soma dos três valores.

In [ ]:
#Solicita os números ao usuário e atribui às variáveis:
v1 = int(input('Digite o primeiro número: '))
v2 = int(input('Digite o segundo número: '))
v3 = int(input('Agora, digite terceiro número: '))

soma = v1 + v2 + v3 #Realiza a soma das variáveis
print(f'A soma de {v1} , {v2} e {v3} é: {soma}')#Imprime os resultados

Digite o primeiro número: 10
Digite o segundo número: 5
Agora, digite terceiro número: 6
A soma de 10 , 5 e 6 é: 21


#### Questão 3
Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a subtração do primeiro pelo o segundo valor.

In [ ]:
#Solicita os números ao usuário e atribui às variáveis:
v1 = int(input('Digite o primeiro número: '))
v2 = int(input('Digite o segundo número: '))
sub = v1 - v2 #Realiza o cálculo
print(f'A subtração de {v1} e {v2} é: {sub}')#Imprime o resultado


Digite o primeiro número: 10
Digite o segundo número: 5
A subtração de 10 e 5 é 2 é: 5


#### Questão 4

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a multiplicação dos dois valores.


In [ ]:
n1 = int(input('Digite o primeiro número: '))
n2 = int(input('Digite o segundo número: '))
mult = n1*n2
print(f'A multiplicação de {n1} por {n2} resultará {mult}')


Digite o primeiro número: 56
Digite o segundo número: 89
A multiplicação de 56 por 89 resultará 4984


#### Questão 5

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e realize a divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.


In [ ]:
num = int(input('Digite um numerador: '))
den = int(input('Digite um denominador (não pode ser 0): '))
div = num/den
print (f'A divisão de {num} por {den} resultará {div:.2f}')

Digite um numerador: 20
Digite um denominador (não pode ser 0): 36
A divisão de 20 por 36 resultará 0.56


#### Questão 6

Crie um programa que solicite dois valores numéricos, um operador e uma potência, e realize a exponenciação entre esses dois valores.


In [ ]:
operador = int(input('Digite um número a ser elevado: '))
potencia = int(input('Digite o número que será a potência: '))
resultado = operador ** potencia
print(f'O resultado de {operador} elevado a {potencia} será de {resultado:.2f}')


Digite um número a ser elevado: 3
Digite o número que será a potência: 3
O resultado de 3 elevado a 3 será de 27.00


#### Questão 7
Crie um programa que solicite dois valores numéricos, um numerador e um denominador e realize a divisão inteira entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [ ]:
num = int(input('Digite um numerador: '))
den = int(input('Digite um denominador (não pode ser 0): '))
div = num//den
print (f'A divisão inteira de {num} por {den} resultará {div:.2f}')

Digite um numerador: 9
Digite um denominador (não pode ser 0): 5
A divisão inteira de 9 por 5 resultará 1.00


#### Questão 8

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e retorne o resto da divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [ ]:
num = int(input('Digite um numerador: '))
den = int(input('Digite um denominador (não pode ser 0): '))
div = num%den
print (f'O resto da divisão inteira de {num} por {den} resultará {div:.2f}')

Digite um numerador: 10
Digite um denominador (não pode ser 0): 3
O resto da divisão inteira de 10 por 3 resultará 1.00


#### Questão 9

Crie um código que solicita 3 notas de um estudante e imprima a média das notas.

In [ ]:
print('***Bem-vindo a calculadora de médias***')
print('***Utilize "." e não "," para notas quebradas***')
aluno = input('Digite o nome do aluno: ')
n1 = float(input('Digite a primeira nota: '))
n2 = float(input('Digite a segunda nota: '))
n3 = float(input('Digite a terceira nota: '))
media = (n1+n2+n3)/3
print(f'A média do aluno {aluno} é de {media:.2f}')


***Bem-vindo a calculadora de médias***
***Utilize "." e não "," para notas quebradas***
Digite o nome do aluno: José
Digite a primeira nota: 6
Digite a segunda nota: 7
Digite a terceira nota: 5
A média do aluno: José é de 6.00


#### Questão 10

Crie um código que calcule e imprima a média ponderada dos números 5, 12, 20 e 15 com pesos respectivamente iguais a 1, 2, 3 e 4.

In [ ]:
n1 = 5*1
n2 = 12*2
n3 = 20*3
n4 = 15*4
ponderada = n1+n2+n3+n4/8
print(f'A média ponderada dos números 5, 12, 20 e 15 com pesos respectivamente iguais a 1, 2, 3 e 4 é de {ponderada}')

A média ponderada dos números 5, 12, 20 e 15 com pesos respectivamente iguais a 1, 2, 3 e 4 é de 96.5


### Editando textos

#### Questão 1

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase na tela.

In [ ]:
frase = 'Olá mundo!'
print(frase)

Olá mundo!


#### Questão 2
Crie um código que solicite uma frase e depois imprima a frase na tela.

In [ ]:
frase = str(input('Digite a sua frase: '))
print(frase)

Digite a sua frase: Olá, como vai?
Olá, como vai?


#### Questão 3

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras maiúsculas.

In [ ]:
frase = str(input('Digite a sua frase: '))
print(frase.upper())

Digite a sua frase: dadwdwd
DADWDWD


#### Questão 4

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras minúsculas.

In [ ]:
frase = str(input('Digite a sua frase: '))
print(frase.lower())

Digite a sua frase: ADUWDHAUD
aduwdhaud


#### Questão 5

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase sem espaços em branco no início e no fim.

In [ ]:
frase = ' Teste para excluir os espaços desta string :)   '
print(frase.strip())

Teste para excluir os espaços desta string :)


#### Questão 6

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim.

In [ ]:
frase = input('Digite a sua frase: ')
print(frase.strip())

Digite a sua frase:       olá     
olá


#### Questão 7


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim e em letras minúsculas.

In [ ]:
frase = input('Digite a sua frase: ')
print(frase.strip().upper())

Digite a sua frase: como vai?
COMO VAI?


#### Questão 8

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`e`” trocadas pela letra “`f`”.

In [ ]:
frase = input('Digite a sua frase: ')
print(frase.replace('e', 'f'))

Digite a sua frase:     sente-se
    sfntf-sf


#### Questão 9


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`a`” trocadas pela caractere  “`@`”.

In [ ]:
frase = input('Digite a sua frase: ')
print(frase.replace('a', '@'))

Digite a sua frase: Quantos anos você tem?
Qu@ntos @nos você tem?


#### Questão 10


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as consoantes “`s`” trocadas pelo caractere  “`$`”.

In [ ]:
frase = input('Digite a sua frase: ')
print(frase.replace('s', '$'))

Digite a sua frase: Sinta-se a vontade sorrindo
Sinta-$e a vontade $orrindo


## Aula 3 - Estruturas condicionais

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça à pessoa usuária para fornecer dois números e exibir o número maior.

In [ ]:
number_1 = float(input('Digite um número: '))
number_2 = float(input('Digite outro número: '))
if number_1 > number_2:
  print(f'O {number_1} é maior que o {number_2}')
elif number_1 == number_2:
  print('São números iguas!')
else:
  print(f'O {number_2} é maior que o {number_1}')

Digite um número: 11
Digite outro número: 23
O 23.0 é maior que o 11.0


#### Questão 2
Escreva um programa que solicite o percentual de crescimento de produção de uma empresa e informe se houve um crescimento (porcentagem positiva) ou decrescimento (porcentagem negativa).

In [ ]:
variacao = float(input('Digite o percentual de crescimento da produção: '))
if variacao > 0:
  print(f'O crescimento foi positivo, de {variacao}%!')
elif variacao == 0:
  print('Não houve crescimento nem decrescimento!')
else:
  print(f'Houve descrescimento de {variacao}%!')

Digite o percentual de crescimento da produção: -9
Houve descrescimento de -9.0%!


#### Questão 3
Escreva um programa que determine se uma letra fornecida pela pessoa usuária é uma vogal ou consoante.

In [ ]:
#Solicita uma letra ao usuário, transformando-a em maiúscula:
letra = input('Digite uma letra: ').upper()
vogais = 'AEIOU'

#Verifica se a letra é vogal ou consoante: e imprime o resultado:
if letra in vogais:
  print (f'A letra "{letra}" é uma vogal!')
else:
  print (f'A letra "{letra}" é uma consoante!')

Digite uma letra: a
A letra "A" é uma vogal!


#### Questão 4
Escreva um programa que leia valores médios de preços de um modelo de carro por 3 anos consecutivos e exiba o valor mais alto e mais baixo entre esses três anos.

In [ ]:
#Solicita ao usuário o preço médio dos últimos 3 anos:
price_year1 = float(input('Informe o preco do veículo no primeiro ano: R$ '))
price_year2 = float(input('Informe o preco do veículo no segundo ano: R$ '))
price_year3 = float(input('Informe o preco do veículo no terceiro ano: R$ '))

#Verifica o maior e menor preço dos valores informados pelo usuário:
higher_price = price_year1
if price_year2 > higher_price:
  higher_price = price_year2
elif price_year3 > higher_price:
  higher_price = price_year3
lower_price = price_year1
if price_year2 < lower_price:
  lower_price = price_year2
elif price_year3 < lower_price:
  lower_price = price_year3

#Imprime o maior e o menor valor informados pelo usuário:
print(f'O maior valor foi de R${higher_price} e o menor de R${lower_price} ')



Informe o preco do veículo no primeiro ano: R$ 10000
Informe o preco do veículo no segundo ano: R$ 30000
Informe o preco do veículo no terceiro ano: R$ 250000
O maior valor foi de R$30000.0 e o menor de R$10000.0 


#### Questão 5
Escreva um programa que pergunte sobre o preço de três produtos e indique qual é o produto mais barato para comprar.

In [ ]:
#Solicita ao usuário o preço dos 3 produtos:
product_1 = float(input(f'Digite o valor do produto 1: R$ '))
product_2 = float(input(f'Digite o valor do produto 2: R$ '))
product_3 = float(input(f'Digite o valor do produto 3: R$ '))

#Verifica o menor preço entre os produtos informados:
lower_price = product_1
if product_2 < lower_price:
  lower_price = product_2
elif product_3 < lower_price:
  lower_price = product_3
elif product_1 == product_2 == product_3:
  print('Todos os preços são idênticos!')

#Imprime o menor valor dentre eles:
print(f'O menor preço é R${lower_price}')


Digite o valor do produto 1: R$ 5
Digite o valor do produto 2: R$ 5
Digite o valor do produto 3: R$ 5
Todos os preços são idênticos!
O menor preço é R$5.0


#### Questão 6
Escreva um programa que leia três números e os exiba em ordem decrescente.

In [ ]:
#Solicita 3 números ao usuário:
number_1 = float(input('Digite o primero número: '))
number_2 = float(input('Digite o segundo número: '))
number_3 = float(input('Digite o terceiro número: '))

#Verifica a ordem dos números informados e imprime a sequência em ordem decrescente:
if (number_1 >= number_2) and (number_1 >=number_3):
  print(number_1)
  if number_2 >=number_3:
    print(number_2)
    print(number_3)
  else:
    print(number_3)
    print(number_2)
elif (number_2 >= number_1) and (number_2 >= number_3):
  print(number_2)
  if number_1 >= number_3:
    print(number_1)
    print(number_3)
  else:
    print(number_3)
    print(number_1)
else:
  print (number_3)
  if number_1 >= number_2:
    print (number_1)
    print (number_2)
  else:
    print(number_2)
    print(number_1)


Digite o primero número: 10
Digite o segundo número: 3
Digite o terceiro número: 23
23.0
10.0
3.0


#### Questão 7
Escreva um programa que pergunte em qual turno a pessoa usuária estuda ("manhã", "tarde" ou "noite") e exiba a mensagem "Bom Dia!", "Boa Tarde!", "Boa Noite!", ou "Valor Inválido!", conforme o caso.

In [ ]:
#Pergunta ao usuário qual é o seu turno de estudo:
turno = input('Informe o turno em que você estuda: Manhã, tarde ou noite?').lower()

#Verifica o turno e imprime a resposta correspondente:
if turno == 'manhã':
  print('Bom dia!')
elif turno == 'tarde':
  print('Boa tarde!')
elif turno == 'noite':
  print('Boa noite!')
else:
  print('Resposta inválida.')


Informe o turno em que você estuda: Manhã, tarde ou noite?TARDE
Boa tarde!


#### Questão 8
Escreva um programa que peça um número inteiro à pessoa usuária e determine se ele é par ou ímpar. Dica: Você pode utilizar o operador módulo `%`.

In [ ]:
"""
Solicita um número ao usuário e verifica o resto da divisão, imprimindo se o
 valor digitado é par ou ímpar.
"""
print(' *** VERIFICADOR: PAR / ÍMPAR *** ')
number = int(input('Digite um número inteiro: '))
if number%2 == 0:
  print(f'O número {number} é par!')
else:
  print(f'O número {number} é ímpar!')

 *** VERIFICADOR: PAR / ÍMPAR *** 
Digite um número inteiro: 23
O número 23 é ímpar!


#### Questão 9
Escreva um programa que peça um número à pessoa usuária e informe se ele é inteiro ou decimal.

In [ ]:
#Solicita um número ao usuário:
num = float(input('Digite um número: '))

# Verifica o resto da divisão para determinar se é inteiro ou decimal:
if num % 1 == 0:
    print('O número é inteiro.')
else:
    print('O número é decimal.')

Digite um número: 10


### Momento dos projetos

#### Questão 10
Um programa deve ser escrito para ler dois números e, em seguida, perguntar à pessoa usuária qual operação ele deseja realizar. O resultado da operação deve incluir informações sobre o número - se é par ou ímpar, positivo ou negativo e inteiro ou decimal.

In [ ]:
#Solicita as entradas ao usuário: números e qual a operação desejada:
number_1 = float(input('Digite o primeiro número: '))
number_2 = float(input('Digite o segundo número: '))
operation = input('Qual operação vai realizar? Digite: "+", "-", "*", ou "/"')

#Verifica a operação escolhida e realiza seu cálculo e o imprime:
if operation == "+":
  result = number_1 + number_2
elif operation == "-":
  result = number_1 - number_2
elif operation == "*":
  result = number_1 * number_2
elif operation == "/":
  result = number_1 / number_2
else:
  resultado = 0
  print ('Operação inválida, o resultado será 0.')
print(f'O resultado da operação solicitada será: {result}')

#Faz as verificações: par ou ímpar; positivo ou negativo; inteiro ou decimal:
if result % 2 == 0:
  print(f'{result} é par!')
else:
  print(f'{result} é ímpar!')
if result >= 0:
  print(f'{result} é positivo!')
else:
  print(f'{result} é negativo!')
if result % 1 == 0:
  print(f'{result} é um número inteiro')
else:
  print(f'{result} é um número decimal')




Digite o primeiro número: 15
Digite o segundo número: 98
Qual operação vai realizar? Digite: "+", "-", "*", ou "/"/
O resultado da operação solicitada será: 0.15306122448979592
0.15306122448979592 é ímpar!
0.15306122448979592 é positivo!
0.15306122448979592 é um número decimal


#### Questão 11
Escreva um programa que peça à pessoa usuária três números que representam os lados de um triângulo. O programa deve informar se os valores podem ser utilizados para formar um triângulo e, caso afirmativo, se ele é equilátero, isósceles ou escaleno. Tenha em mente algumas dicas:
- Três lados formam um triângulo quando a soma de quaisquer dois lados for maior que o terceiro;
- Triângulo Equilátero: três lados iguais;
- Triângulo Isósceles: quaisquer dois lados iguais;
- Triângulo Escaleno: três lados diferentes;


In [ ]:
# Coletamos os lados de um triângulo
print('Coletaremos os lados de um triângulo.')
lado1 = float(input('Digite o comprimento do primeiro lado: '))
lado2 = float(input('Digite o comprimento do segundo lado: '))
lado3 = float(input('Digite o comprimento do terceiro lado: '))

# Verificamos de os lados formam um triângulo
if (lado1 + lado2 > lado3) and (lado2 + lado3 > lado1) and (lado1 + lado3 > lado2):
    print('Os valores podem formar um triângulo!')

# comparamos os lados para verificar o tipo de triângulo
    if (lado1 == lado2) and (lado2 == lado3):
        print('O triângulo é equilátero.')
    elif (lado1 != lado2) and (lado2 != lado3) and (lado1 != lado3):
        print('O triângulo é escaleno.')
    else:
        print('O triângulo é isósceles.')
else:
    print('Os valores não podem formar um triângulo!')

Coletaremos os lados de um triângulo.
Digite o comprimento do primeiro lado: 6
Digite o comprimento do segundo lado: 6
Digite o comprimento do terceiro lado: 5
Os valores podem formar um triângulo!
O triângulo é isósceles.


#### Questão 12
Um estabelecimento está vendendo combustíveis com descontos variados. Para o etanol, se a quantidade comprada for até 15 litros, o desconto será de 2% por litro. Caso contrário, será de 4% por litro. Para o diesel, se a quantidade comprada for até 15 litros, o desconto será de 3% por litro. Caso contrário, será de 5% por litro. O preço do litro de diesel é R\$ 2,00 e o preço do litro de etanol é R\$ 1,70. Escreva um programa que leia a quantidade de litros vendidos e o tipo de combustível (E para etanol e D para diesel) e calcule o valor a ser pago pelo cliente. Tenha em mente algumas dicas:

- O do valor do desconto será a multiplicação entre preço do litro, quantidade de litros e o valor do desconto.
- O valor a ser pago por um cliente será o resultado da multiplicação do preço do litro pela quantidade de litros menos o valor de desconto resultante do cálculo.


In [ ]:
#Solicita as entradas do usuário e declara as variáveis:
quantidade = float(input('Quantidade de litros: '))
tipo = str(input('Qual é o tipo de combustível? Diesel (D) | Etanol (E)')).upper()
preço_diesel = 2.00
preço_etanol = 1.70
desconto = 0
valor = 0

#Calcula o desconto com base no tipo e quantidade de combustível:
if quantidade <= 15 and tipo == 'E':
  desconto = (preço_etanol * quantidade)* 0.02
elif quantidade >= 15 and tipo == 'E':
  desconto = (preço_etanol * quantidade)* 0.04
elif quantidade <= 15 and tipo == 'D':
  desconto = (preço_diesel * quantidade)* 0.03
elif quantidade >= 15 and tipo == 'D':
  desconto = (preço_diesel * quantidade)* 0.05

#Calcula o valor com base no tipo e quantidade de combustível:
if tipo  == 'E':
  valor = quantidade * preço_etanol
elif tipo == 'D':
  valor = quantidade * preço_diesel

#Define o resultado com base no valor - desconto e retorna ao usuário:
resultado = valor - desconto
print(f'No total, você abasteceu {quantidade} litros, com R${desconto:.2f} de desconto')
print(f'Totalizando o valor de R${resultado}')




Quantidade de litros: 120
Qual é o tipo de combustível? Diesel (D) | Etanol (E)d
No total, você abasteceu 120.0 litros, com R$12.00 de desconto
Totalizando o valor de R$ 228.0


#### Questão 13
Em uma empresa de venda de imóveis você precisa criar um código que analise os dados de vendas anuais para ajudar a diretoria na tomada de decisão. O código precisa coletar os dados de quantidade de venda durante os anos de 2022 e 2023 e fazer um cálculo de variação percentual. A partir do valor da variação, deve ser enviada às seguintes sugestões:
- Para variação acima de 20%: bonificação para o time de vendas.
- Para variação entre 2% e 20%: pequena bonificação para time de vendas.
- Para variação entre 2% e -10%: planejamento de políticas de incentivo às vendas.
- Para bonificações abaixo de -10%: corte de gastos.


In [ ]:
#Coleta as entradas do usuário e define as variáveis:
vendas_2022 = float(input('Quantidade de vendas de 2022: '))
vendas_2023 = float(input('Quantidade de vendas de 2023: '))
var_percentual = 100 * (vendas_2023 - vendas_2022) / (vendas_2022)

#Calcula a variação percentual e retorna o resultado ao usuário:
if var_percentual > 20:
  print(f'Variação de {var_percentual:.2f}%: bonificação para o time de vendas!')
elif 2 <= var_percentual <= 20:
  print(f'Variação de {var_percentual:.2f}%: pequena bonificação para o time de vendas!')
elif -10 <= var_percentual <= 2:
  print(f'Variação de {var_percentual:.2f}%:  planejamento de políticas de incentivo às vendas!')
else:
  print(f'Variação de {var_percentual:.2f}%:  Corte de gastos!')




Quantidade de vendas de 2022: 100
Quantidade de vendas de 2023: 59595
Variação de 59495.00%: bonificação para o time de vendas!


## Aula 4 - Estruturas de repetição

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça dois números inteiros e imprima todos os números inteiros entre eles.

In [ ]:
#Solicita as entradas ao usuário:
inicio = int(input('Digite o número para início:'))
fim = int(input('Digite o número para finalizar:'))

#Verifica o maior entre início e fim e imprime os números inteiros entre eles:
if inicio < fim:
  for indice in range(inicio + 1, fim):
    print(indice)
elif fim > inicio:
  for indice in range(fim + 1, inicio):
    print(indice)
else:
  print('Os números digitados são idênticos.')


Digite o número para início:5
Digite o número para finalizar:23
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22


#### Questão 2
Escreva um programa para calcular quantos dias levará para a colônia de uma bactéria A ultrapassar ou igualar a colônia de uma bactéria B, com base nas taxas de crescimento de 3% e 1,5% respectivamente. Considere que a colônia A inicia com 4 elementos e a colônia B com 10 elementos.

In [ ]:
#Guarda as variáveis da quantidade, taxas e dias:
bacteria_a = 4
bacteria_b = 10

taxa_a = 0.03
taxa_b = 0.015

dias = 0

#Faz o cálculo multiplicando com base na taxa de cada um e imprime o resultado:
while bacteria_a < bacteria_b:
  bacteria_a *= 1 + taxa_a
  bacteria_b *= 1 + taxa_b
  dias += 1
print(f'A colônia da bactéria A irá levar {dias} dias para ultrapassar a colônia da bactéria B')
print(f'Quando a colônia A terá {bacteria_a:.2f} elementos e a colônia B {bacteria_b:.2f}.')




A colônia da bactéria A irá levar 63 dias para ultrapassar a colônia da bactéria B
Quando a colônia A terá 25.75 elementos e a colônia B 25.55.


#### Questão 3
Para tratar uma quantidade de 15 dados de avaliações de pessoas usuárias de um serviço da empresa, precisamos verificar se as notas são válidas. Então, escreva um programa que vai receber a nota de 0 a 5 de todos os dados e verificar se é um valor válido. Caso seja inserido uma nota acima de 5 ou abaixo de 0, repita até que a pessoa usuária insira um valor válido.

In [ ]:
#Inicia um laço para a variável "i" no range 1, 16 (para coletar do usuário 1 a 15:
for i in range(1, 16):
  nota = float(input(f'Digite a nota da pessoa usuária {i}: '))

#Verifica se a nota atende ao requisito 1 a 5, se não, pede uma nota válida:
  if nota < 0 or nota > 5:
    nota = float(input(f'Nota inválida, digite uma nota de 0 a 5! Usuário {i}: '))

#Continua no laço até o fim do range:
    continue
print(f'Fim da veriricação, todas as notas estão entre 0 e 5!')


Digite a nota da pessoa usuária 1: 5
Digite a nota da pessoa usuária 2: 3
Digite a nota da pessoa usuária 3: 3
Digite a nota da pessoa usuária 4: 3
Digite a nota da pessoa usuária 5: 3
Digite a nota da pessoa usuária 6: 10
Nota inválida, digite uma nota de 0 a 5! Usuário 6: 3
Digite a nota da pessoa usuária 7: 3
Digite a nota da pessoa usuária 8: 3
Digite a nota da pessoa usuária 9: 3
Digite a nota da pessoa usuária 10: 3
Digite a nota da pessoa usuária 11: 3
Digite a nota da pessoa usuária 12: 3
Digite a nota da pessoa usuária 13: 3
Digite a nota da pessoa usuária 14: 3
Digite a nota da pessoa usuária 15: 3
Fim da veriricação, todas as notas estão entre 0 e 5!


#### Questão 4
Desenvolva um programa que leia um conjunto indeterminado de temperaturas em Celsius e faça uma análise. Portanto, escreva um programa que leia temperaturas e informe a média delas. A leitura deve ser encerrada ao ser enviado o valor -273°C.

In [ ]:
temperatura = float(input('Digite a temperatura em Celsius: '))
contadora = 0
soma = 0

while temperatura != -273:
  soma += temperatura
  contadora += 1
  temperatura = float(input('Insira a temperatura em Celsius: '))

media = soma / contadora
print(f'A média das temperaturas é: {media}')

Digite a temperatura em Celsius: 20
Insira a temperatura em Celsius: 23
Insira a temperatura em Celsius: 25
Insira a temperatura em Celsius: 23
Insira a temperatura em Celsius: 23
Insira a temperatura em Celsius: 23
Insira a temperatura em Celsius: -5895858
Insira a temperatura em Celsius: -595959


KeyboardInterrupt: ignored

#### Questão 5
Escreva um programa que calcule o fatorial de um número inteiro fornecido pela pessoa usuária. O fatorial de um número inteiro é a multiplicação desse número por todos os seus antecessores até o número 1. Por exemplo, o fatorial de 5 é 5 x 4 x 3 x 2 x 1 = 120.

In [ ]:
numero = int (input ('Digite um número: '))
fatorial = 1

for contador in range(1, numero + 1):
    fatorial *= contador

print(f'O fatorial de {numero} é {fatorial:.2f}')

Digite um número: 10
O fatorial de 10 é 3628800.00


### Momento dos projetos

#### Questão 6
Escreva um programa que gere a tabuada de um número inteiro de 1 a 10, de acordo com a escolha da pessoa usuária. Como exemplo, para o número 2, a tabuada deve ser mostrada no seguinte formato:
```
Tabuada do 2:
2 x 1 = 2
2 x 2 = 4
[...]
2 x 10 = 20
```

In [ ]:
num = int(input('Insira um número para a taboada: '))

print(f'Taboada do número {num}')
for i in range (1,11):
  resultado = num * i
  print(f' {num} x {i} = {resultado}')


Insira um número para a taboada: 10
Taboada do número 10
 10 x 1 = 10
 10 x 2 = 20
 10 x 3 = 30
 10 x 4 = 40
 10 x 5 = 50
 10 x 6 = 60
 10 x 7 = 70
 10 x 8 = 80
 10 x 9 = 90
 10 x 10 = 100


#### Questão 7
Os números primos possuem várias aplicações dentro da Ciência de Dados, por exemplo, na criptografia e segurança. Um número primo é aquele que é divisível apenas por um e por ele mesmo. Faça um programa que peça um número inteiro e determine se ele é ou não um número primo.

In [ ]:
num = int(input('Insira um número para saber se é primo ou não: '))
if num > 1:
  for i in range (2, num):
    if  (num % i) == 0:
      print(f'O número {num} não é primo')
      break
    else:
      print(f'O número {num} é primo.')
else:
  print(f'O número {num} não é primo.')

Insira um número para saber se é primo ou não: 10
O número 10 não é primo


#### Questão 8
Vamos entender a distribuição de idades de pensionistas de uma empresa de previdência. Escreva um programa que leia as idades de uma quantidade não informada de clientes e mostre a distribuição em intervalos de [0-25], [26-50], [51-75] e [76-100]. Encerre a entrada de dados com um número negativo.

In [ ]:
#Solicita ao usuário a idade:
age = int(input('Digite a idade ou um número negativo para encerrar: '))

#Variáveis armazenam de acordo com a faixa etária:
contador_0_25 = 0
contador_26_50 = 0
contador_51_75 = 0
contador_76_100 = 0

#Laço armazena os inputs de acordo com as idades:
while age > 0:
  if age >=0 and age <=25:
    contador_0_25 += 1
  elif age >=26 and age <=50:
    contador_26_50 += 1
  elif age >=51 and age <=75:
    contador_51_75 += 1
  elif age >=76 and age <=100:
    contador_76_100 += 1

#Laço continua solicitando a entrada até que o usuário digite um número negativo para encerrar:
  age = int(input('Digite a idade ou um número negativo para encerrar: '))

#Imprime o resultado da distribuição de idades por faixa etária:
print('Distribuição de idades:')
print('[0-25]:', contador_0_25)
print('[26-50]:', contador_26_50)
print('[51-75]:', contador_51_75)
print('[76-100]:', contador_76_100)





Digite a idade ou um número negativo para encerrar: 56
Digite a idade ou um número negativo para encerrar: 31
Digite a idade ou um número negativo para encerrar: 25
Digite a idade ou um número negativo para encerrar: 20
Digite a idade ou um número negativo para encerrar: 15
Digite a idade ou um número negativo para encerrar: 17
Digite a idade ou um número negativo para encerrar: 75
Digite a idade ou um número negativo para encerrar: 66
Digite a idade ou um número negativo para encerrar: 89
Digite a idade ou um número negativo para encerrar: 52
Digite a idade ou um número negativo para encerrar: 44
Digite a idade ou um número negativo para encerrar: -6
Distribuição de idades:
[0-25]: 4
[26-50]: 2
[51-75]: 4
[76-100]: 1


#### Questão 9
Em uma eleição para gerência em uma empresa com 20 funcionários, existem quatro candidatos. Escreva um programa que calcule o vencedor da eleição. A votação ocorreu da seguinte maneira:
- Cada funcionário votou em um dos quatro candidatos (representados pelos números 1, 2, 3 e 4).
- Também foram contabilizados os votos nulos (representado pelo número 5) e os votos em branco (representado pelo número 6).

Ao final da votação, o programa deve exibir o total de votos para cada candidato, o número de votos nulos e o número de votos em branco. Além disso, deve calcular e exibir a porcentagem de votos nulos em relação ao total de votos e a porcentagem de votos em branco em relação ao total de votos.

In [ ]:
#Declara as variavéis para armazenar os votos:
votos_candidato1 = 0
votos_candidato2 = 0
votos_candidato3 = 0
votos_candidato4 = 0
votos_nulos = 0
votos_branco = 0

#Cria um laço para receber os votos e armazená-los:
for i in range(0,20):
    voto = int(input('Informe seu voto: '))
    if voto == 1:
        votos_candidato1 += 1
    elif voto == 2:
        votos_candidato2 += 1
    elif voto == 3:
        votos_candidato3 += 1
    elif voto == 4:
        votos_candidato4 += 1
    elif voto == 5:
        votos_nulos += 1
    elif voto == 6:
        votos_branco += 1
    else:
        print("Voto inválido.")

#Imprime o resultado:
print(f'Votos candidato 1: {votos_candidato1}')
print(f'Votos candidato 2: {votos_candidato2}')
print(f'Votos candidato 3: {votos_candidato3}')
print(f'Votos candidato 4: {votos_candidato4}')
print(f'Votos nulos: {votos_nulos}')
print(f'Votos em branco: {votos_branco}')
print(f'Percentual de votos nulos: {(votos_nulos / 20 * 100)}%')
print(f'Percentual de votos em branco: {(votos_branco / 20 * 100)}%')

Informe seu voto: 1
Informe seu voto: 2
Informe seu voto: 3
Informe seu voto: 2
Informe seu voto: 5
Informe seu voto: 3
Informe seu voto: 6
Informe seu voto: 6
Informe seu voto: 5
Informe seu voto: 1
Informe seu voto: 2
Informe seu voto: 1
Informe seu voto: 3
Informe seu voto: 4
Informe seu voto: 5
Informe seu voto: 6
Informe seu voto: 4
Informe seu voto: 2
Informe seu voto: 3
Informe seu voto: 1
Votos candidato 1: 4
Votos candidato 2: 4
Votos candidato 3: 4
Votos candidato 4: 2
Votos nulos: 3
Votos em branco: 3
Percentual de votos nulos: 15.0
Percentual de votos em branco: 15.0


## Aula 5 - Estruturas de dados

### Aquecendo na programação

#### Questão 1
Faça um programa que tenha a seguinte lista contendo os valores de gastos de uma empresa de papel `[2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]`. Com esses valores, faça um programa que calcule a média de gastos. Dica, use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).

In [ ]:
gastos = [2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]
soma = sum(gastos)
quantidade = len(gastos)
mediagastos = soma/quantidade
print(f'A média de gastos foi de {mediagastos:.2f} reais')



A média de gastos foi de 3144.22 reais


#### Questão 2
 Com os mesmos dados da questão anterior, defina quantas compras foram acima de 3000 reais e calcule a porcentagem quanto ao total de compras.

In [ ]:
gastos = [2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]

up3000 = 0

for contador in gastos:
  if contador > 3000:
    up3000 += 1

porcetagem_up3000 = 100* (up3000) / len(gastos)

print(f'{up3000} compras foram acima de R$3000,00.')
print(f'{porcetagem_up3000}% dos gastos foram acima de R$3000,00.')


7 compras foram acima de R$3000,00.
70.0% dos gastos foram acima de R$3000,00.


#### Questão 3
Faça um código que colete em uma lista 5 números inteiros quaisquer e imprima a lista. Exemplo: `[1,4,7,2,4]`.

In [ ]:
#Variável recebe uma lista vazia:
lista_numeros = []
#Laço com range 0 a 5 para coletar 5 números e insere ao final da lista com o método append:
for i in range (0, 5):
  numero = int(input('Digite um número inteiro:'))
  lista_numeros.append(numero)
#Imprime a lista:
print(f'Lista de números inseridos: {lista_numeros}')

Digite um número inteiro:6
Digite um número inteiro:5
Digite um número inteiro:3
Digite um número inteiro:4
Digite um número inteiro:9
Lista de números inseridos: [6, 5, 3, 4, 9]


#### Questão 4
Colete novamente 5 inteiros e imprima a lista em ordem inversa à enviada.

In [ ]:
#Variável recebe uma lista vazia:
lista_numeros = []

#Laço com range 0 a 5 para coletar 5 números e insere ao final da lista com o método append:
for i in range (0, 5):
  numero = int(input('Digite um número inteiro:'))
  lista_numeros.append(numero)

#O método reverse() irá inverter a ordem digitada pelo usuário:
lista_numeros.reverse()

#Imprime o resultado:
print(f'Lista de números inseridos: {lista_numeros}')

Digite um número inteiro:52
Digite um número inteiro:8
Digite um número inteiro:36
Digite um número inteiro:21
Digite um número inteiro:5
Lista de números inseridos: [5, 21, 36, 8, 52]


#### Questão 5
Faça um programa que, ao inserir um número qualquer, criará uma lista contendo todos os números primos entre 1 e o número digitado.

In [ ]:

numero = int(input('Digite um número inteiro: '))
lista_primos = []

for num in range(2, numero):
  primo = True
  for divisiveis in range (2, num):
    if num % divisiveis == 0:
      primo = False
      break
  if primo:
    lista_primos.append(num)
print(f'Lista de números primos: {lista_primos}')






Digite um número inteiro: 60
Lista de números primos: [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59]


#### Questão 6
Escreva um programa que peça uma data informando o dia, mês e ano e determine se ela é uma data válida para uma análise.

In [ ]:
dia = int(input('Digite o dia: '))
mes = int(input('Digite o mês: '))
ano = int(input('Digite o ano: '))

if mes == 2:
  # Verificamos se é ou não um ano bissexto
  if ano % 4 == 0 and (ano % 400 == 0 or ano % 100 != 0):
    dias_fevereiro = 29
  else:
    dias_fevereiro = 28
  # Verificamos se o dia colocado corresponde ao máximo de dias de fevereiro
  if dia >= 1 and dia <= dias_fevereiro:
    print('Data válida')
  else:
    print('Data inválida')
# Verificamos meses terminados em 31 dias
elif mes in [1, 3, 5, 7, 8, 10, 12]:
  if dia >= 1 and dia <= 31:
    print('Data válida')
  else:
    print('Data inválida')
# Verificamos meses terminados em 30 dias
elif mes in [4, 6, 9, 11]:
  if dia >= 1 and dia <= 30:
    print('Data válida')
  else:
    print('Data inválida')
# Caso o mês não esteja entre 1 e 12
else:
  print('Data inválida')

Digite o dia: 23
Digite o mês: 6
Digite o ano: 2013
Data válida


### Momento dos projetos

#### Questão 7
Para um estudo envolvendo o nível de multiplicação de bactérias em uma colônia, foi coletado o número de bactérias multiplicadas por dia, com base em um processo de duplicação diária e pode ser observado a seguir: `[1.2, 2.1, 3.3, 5.0, 7.8, 11.3, 16.6, 25.1, 37.8, 56.9]`. Tendo esses valores, faça um código que gere uma lista contendo o percentual de crescimento de bactérias por dia, comparando o número de bactérias em cada dia com o número de bactérias do dia anterior. Dica: para calcular o percentual de crescimento usamos a seguinte equação: `100 * (amostra_atual - amostra_passada) / (amostra_passada)`.

In [ ]:
#Variável armazena a lista da colônia de bactérias:
colonia_bacterias = [1.2, 2.1, 3.3, 5.0, 7.8, 11.3, 16.6, 25.1, 37.8, 56.9]

#Lista vazia contendo o percentual de crescimento:
percentual_crescimento = []

#laço para iterar sobre a variável "colonia_bacterias' e calcular a porcentagem:"
for i in range(1, len(colonia_bacterias)):
  porcentagem = 100 * (colonia_bacterias[i] - colonia_bacterias[i - 1]) / (colonia_bacterias[i - 1])
  percentual_crescimento.append(porcentagem)

#Saída dos resultados com laço para formatar cada item da lista:
print('Porcentagens de crescimento:')
for valor in percentual_crescimento:
    print(f'{valor:.2f}%')

Porcentagens de crescimento:
75.00%
57.14%
51.52%
56.00%
44.87%
46.90%
51.20%
50.60%
50.53%


#### Questão 8
Para uma seleção de produtos alimentícios, precisamos separar o conjunto de IDs dados por números inteiros, sabendo que os produtos com ID par são doces e os com ID ímpar são amargos. Monte um código que colete 10 IDs. Depois, calcule e mostre a quantidade de produtos doces e amargos.

In [ ]:
id_doce = []
id_amargo = []

for id in range (0, 10):
  id_produto = int(input('Digite o ID do produto: '))
  if id_produto % 2 == 0:
    id_doce.append(id_produto)
  else:
    id_amargo.append(id_produto)

soma_doce = len(id_doce)
soma_amargo = len(id_amargo)


print(f'Dentre os IDs informados {id_doce} são doces e {id_amargo} são amargos.')
print(f'E somando temos {soma_doce} produtos doces e {soma_amargo} amargos!')



Digite o ID do produto: 1
Digite o ID do produto: 13
Digite o ID do produto: 2
Digite o ID do produto: 14
Digite o ID do produto: 5
Digite o ID do produto: 6
Digite o ID do produto: 9
Digite o ID do produto: 17
Digite o ID do produto: 11
Dentre os IDs informados [2, 14, 6] são doces e [1, 13, 5, 9, 17, 11] são amargos.
E somando temos 3 produtos doces e 6 amargos!


#### Questão 9
Desenvolva um programa que informa a nota de um aluno de acordo com suas respostas. Ele deve pedir a resposta de um aluno para cada questão e é preciso verificar se a resposta foi igual ao gabarito. Cada questão vale um ponto e existem alternativas:  A, B, C ou D.
```
Gabarito da Prova:
01 - D
02 - A
03 - C
04 - B
05 - A
06 - D
07 - C
08 - C
09 - A
10 - B
```

In [ ]:
respostas = []
gabarito = ['D', 'A', 'C', 'B', 'A', 'D', 'C', 'C', 'A', 'B']
nota = 0

for i in range(0, 10):
  respostas.append(input(f'Insira a resposta da questão {i+1}: ').upper())

for i in range(0,10):
  if respostas[i] == gabarito[i]:
    nota += 1
print(f'Nota final {nota}')

Insira a resposta da questão 1: a
Insira a resposta da questão 2: a
Insira a resposta da questão 3: c
Insira a resposta da questão 4: b
Insira a resposta da questão 5: a
Insira a resposta da questão 6: a
Insira a resposta da questão 7: a
Insira a resposta da questão 8: a
Insira a resposta da questão 9: a
Insira a resposta da questão 10: a
Nota final 5


#### Questão 10
Um instituto de meteorologia deseja fazer um estudo de temperatura média de cada mês do ano. Para isso, você precisa fazer um código que colete e armazene essas temperaturas médias em uma lista. Depois, calcule a média anual das temperaturas e mostre todas as temperaturas acima da média anual, e em que mês elas ocorreram, mostrando os meses por extenso: Janeiro, Fevereiro, etc.

In [ ]:
temperaturas_mensais = []
for i in range(0, 12):
  temperaturas_mensais.append(float(input(f'Digite a média de temperatura do mês {i+1}: ')))
meses = ['Janeiro', 'Fevereiro', 'Março', 'Abril', 'Maio', 'Junho', 'Julho', 'Agosto', 'Setembro', 'Outubro', 'Novembro', 'Dezembro']
media_anual = sum(temperaturas_mensais) / len(meses)
print('Temperaturas acima da média em: ')
for i in range(0,12):
  # Verificamos todas as temperaturas de acordo com a média anual
  if temperaturas_mensais[i] > media_anual:
    # Como os índices dos meses correspondem às temperaturas,
    # podemos imprimir eles sob o mesmo índice
    print(meses[i])

Digite a média de temperatura do mês 1: 20
Digite a média de temperatura do mês 2: 23
Digite a média de temperatura do mês 3: 30
Digite a média de temperatura do mês 4: 15
Digite a média de temperatura do mês 5: 32
Digite a média de temperatura do mês 6: 12
Digite a média de temperatura do mês 7: 10
Digite a média de temperatura do mês 8: 23
Digite a média de temperatura do mês 9: 26
Digite a média de temperatura do mês 10: 28
Digite a média de temperatura do mês 11: 33
Digite a média de temperatura do mês 12: 36
Temperaturas acima da média em: 
Março
Maio
Setembro
Outubro
Novembro
Dezembro


#### Questão 11
Uma empresa de e-commerce está interessada em analisar as vendas dos seus produtos. Os dados das vendas foram armazenados em um dicionário:
```
{'Produto A': 300, 'Produto B': 80, 'Produto C': 60,
 'Produto D': 200, 'Produto E': 250, 'Produto F': 30}
```
Escreva um código que calcule o total de vendas e o produto mais vendido.

In [ ]:
dados_vendas = {'Produto A': 300, 'Produto B': 80, 'Produto C': 60,
 'Produto D': 200, 'Produto E': 250, 'Produto F': 30}
total_vendas = 0
p_mais_vendido = ''
unidades_p_mais_vendido = 0

for produto in dados_vendas.keys():
  total_vendas += dados_vendas[produto]
  if dados_vendas[produto] > unidades_p_mais_vendido:
   unidades_p_mais_vendido = dados_vendas[produto]
   p_mais_vendido = produto

print(f'Total de vendas é {total_vendas}')
print(f'{p_mais_vendido} é o mais vendido')



Total de vendas é 920
Produto A é o mais vendido


#### Questão 12
Uma pesquisa de mercado foi feita para decidir qual design de uma marca infantil mais agrada crianças. A pesquisa foi feita e o votos computados podem ser observados abaixo:
```
'''
Tabela de votos da marca
Design 1 - 1334 votos
Design 2 - 982 votos
Design 3 - 1751 votos
Design 4 - 210 votos
Design 5 - 1811 votos
'''
```
Adapte os dados fornecidos a você para uma estrutura de dicionário e a partir dele, informe o design vencedor e a porcentagem de votos recebidos.


In [ ]:
dados_votos = {
    "Design 1": 1334,
    "Design 2": 982,
    "Design 3": 1751,
    "Design 4": 210,
    "Design 5": 1811
}

total_votos = 0
mais_votado = ''
voto_vencedor = 0

# Itera sobre o dicionário de dados de votos
for design, voto_design in dados_votos.items():
    total_votos += voto_design  # Calcula o total de votos

    # Verifica se o design atual tem mais votos que o vencedor atual
    if voto_design > voto_vencedor:
        voto_vencedor = voto_design  # Atualiza o número de votos do vencedor
        mais_votado = design  # Atualiza o design mais votado

# Calcula a porcentagem de votos do design mais votado em relação ao total de votos
porcentagem = (voto_vencedor / total_votos) * 100

# Resultado
print(f'{mais_votado} é o vencedor.')
print(f'Porcentagem de votos: {porcentagem:.2f}%')

Design 5 é o vencedor: 
Porcentagem de votos: 29.75%


#### Questão 13
Os funcionários de um setor da empresa que você trabalha vão receber um abono correspondente a 10% do seu salário devido ao ótimo desempenho do time. O setor financeiro solicitou sua ajuda para a verificação das consequências financeiras que esse abono irá gerar nos recursos. Assim, foi encaminhada para você uma lista com os salários que receberão o abono: `[1172, 1644, 2617, 5130, 5532, 6341, 6650, 7238, 7685, 7782, 7903]`. O abono de cada funcionário não pode ser inferior a 200. Em código, transforme cada um dos salários em chaves de um dicionário e o abono de cada salário no elemento. Depois, informe o total de gastos com o abono, quantos funcionários receberam o abono mínimo e qual o maior valor de abono fornecido.


In [ ]:
# Lista de salários
salarios = [1172, 1644, 2617, 5130, 5532, 6341, 6650, 7238, 7685, 7782, 7903]

dic_abonos = {}
total_abono = 0
abonos_minimo = 0
maior_abono = 0

# Percorremos toda a lista de salários
for salario in salarios:
  # Calculamos o valor teórico de abono
  abono = salario * 0.1
  # Caso o abono seja inferior a 200,
  # ajustamos o valor de abono para o mínimo (200)
  if abono < 200:
    abono = 200
  # Adicionamos um novo dado no dicionário chave abono
  dic_abonos[salario] = abono

# Percorremos todos os valores do dicionário de abonos
for abono in dic_abonos.values():
  # Contamos o salário minimo
  if abono == 200:
    abonos_minimo += 1
  # Verificamos se o abono lido é maior que o valor armazenado em maior_abono
  # Cada vez que o abono superar o valor de maior_abono,
  # a variável maior_abono vai ser igual à abono, atribuindo um novo valor
  if abono > maior_abono:
    maior_abono = abono
  # Somamos os abonos
  total_abono += abono
# Resultados
print(f'Abonos: {dic_abonos}')
print(f'Total de gasto com abonos: {total_abono}')
print(f'Número de funcionários que receberam o abono mínimo: {abonos_minimo}')
print(f'Maior valor de abono: {maior_abono}')


#### Questão 14
Uma equipe de cientistas de dados está estudando a diversidade biológica em uma floresta. A equipe fez a coleta de informações sobre o número de espécies de plantas e animais em cada área da floresta e armazenaram essas informações em um dicionário. Nele, a chave descreve a área dos dados e os valores nas listas correspondem às espécies de plantas e animais nas áreas, respectivamente.
```
{'Área Norte': [2819, 7236],
 'Área Leste': [1440, 9492],
 'Área Sul': [5969, 7496],
 'Área Oeste': [14446, 49688],
 'Área Centro': [22558, 45148]}
```
Escreva um código para calcular a média de espécies por área e identificar a área com a maior diversidade biológica. Dica: use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).


In [ ]:
#Organiza os dados em um dicionário:
dados_floresta = {'Área Norte': [2819, 7236],
 'Área Leste': [1440, 9492],
 'Área Sul': [5969, 7496],
 'Área Oeste': [14446, 49688],
 'Área Centro': [22558, 45148]}

#Declara as variáveis:
soma_media = 0
maior_diverisidade = ''
maior_soma = 0

#Itera sobre o dicionário para calcular soma e média:
for area, especies in dados_floresta.items():
  soma_especies = sum(especies)
  media = soma_especies / len(especies)
  print(f'A {area} tem a média de {media} espécies')

#Verifica se a soma das espécies é maior que o último dado "maior soma e atribui um valor:
  if soma_especies > maior_soma:
      maior_soma = soma_especies
      maior_diversidade = area
      soma_media += media

# A média total será dada pela soma_media dividida pela quantidade de áreas
media_total = soma_media / len(dados_floresta)
print(f'Média geral de espécies: {media_total}')
print(f'Área com a maior diversidade biológica: {maior_diversidade}')



A Área Norte tem a média de 5027.5 espécies
A Área Leste tem a média de 5466.0 espécies
A Área Sul tem a média de 6732.5 espécies
A Área Oeste tem a média de 32067.0 espécies
A Área Centro tem a média de 33853.0 espécies
Média geral de espécies: 16629.2
Área com a maior diversidade biológica: Área Centro


#### Questão 15
O setor de RH da sua empresa te pediu uma ajuda para analisar as idades dos funcionários de 4 setores da empresa. Para isso, ele te forneceu os seguintes dados:
```
{'Setor A': [22, 26, 30, 30, 35, 38, 40, 56, 57, 65],
 'Setor B': [22, 24, 26, 33, 41, 49, 50, 54, 60, 64],
 'Setor C': [23, 26, 26, 29, 34, 35, 36, 41, 52, 56],
 'Setor D': [19, 20, 25, 27, 34, 39, 42, 44, 50, 65]}
```
Sabendo que cada setor tem 10 funcionários, construa um código que calcule a média de idade de cada setor, a idade média geral entre todos os setores e quantas pessoas estão acima da idade média geral.


In [ ]:
# Especificamos os dados para um dicionário
dados = {'Setor A': [22, 26, 30, 30, 35, 38, 40, 56, 57, 65],
        'Setor B': [22, 24, 26, 33, 41, 49, 50, 54, 60, 64],
        'Setor C': [23, 26, 26, 29, 34, 35, 36, 41, 52, 56],
        'Setor D': [19, 20, 25, 27, 34, 39, 42, 44, 50, 65]}
# Inicializamos a variável que irá somar todas as idades
total_idades = 0
# Percorremos os valores de chaves e elementos do dicionário
for setor, idades in dados.items():
  # Calculamos a média dividindo a soma das idades pela quantidade funcionários em cada setor
  media_idade = sum(idades) / len(idades)
  # Imprimimos
  print(f'O {setor} tem a média de {media_idade}')
  # Somamos as médias
  total_idades += sum(idades)
# A média total será dada pela total_idades dividida pela quantidade de pessoas totais (setores * funcionários por setor)
media_total = total_idades / (len(idades) * len(dados))
print(f'A média de idade geral é {media_total}')

# Inicializamos a variável que irá contar todas pessoas com idade acima da média
acima_media = 0
# Percorremos novamente os valores de chaves e elementos do dicionário
for setor, idades in dados.items():
  # Lemos os elementos (idades) dentro de cada lista de idades no dicionário
  for id in idades:
    # Verificamos se o valor da idade é superior à média total
    if id > media_total:
      # Caso o valor da idade seja superior à média, incrementamos mais um no contador
      acima_media += 1
# Resultado
print(f'{acima_media} pessoas estão acima da idade média geral')


O Setor A tem a média de 39.9
O Setor B tem a média de 42.3
O Setor C tem a média de 35.8
O Setor D tem a média de 36.5
A média de idade geral é 38.625
18 pessoas estão acima da idade média geral
